contracts/contracts/token/OUSD.sol

```diff
@@ -445,159 +441,268 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
+        );
+        RebaseOptions stateFrom = rebaseState[_from];
+        RebaseOptions stateTo = rebaseState[_to];
 
-        emit TotalSupplyUpdatedHighres(
-            _totalSupply,
-            _rebasingCredits,
-            _rebasingCreditsPerToken
+        require(
+            stateFrom == RebaseOptions.NotSet ||
+                stateFrom == RebaseOptions.StdNonRebasing ||
+                stateFrom == RebaseOptions.StdRebasing,
+            "Invalid rebaseState from"
+        );
+
+        require(
+            stateTo == RebaseOptions.NotSet ||
+                stateTo == RebaseOptions.StdNonRebasing ||
+                stateTo == RebaseOptions.StdRebasing,
+            "Invalid rebaseState to"
+        );
+
+        if (alternativeCreditsPerToken[_from] == 0) {
+            _rebaseOptOut(_from);
+        }
+        if (alternativeCreditsPerToken[_to] > 0) {
+            _rebaseOptIn(_to);
+        }
+
+        uint256 fromBalance = balanceOf(_from);
+        uint256 toBalance = balanceOf(_to);
+
+        // Set up the bidirectional links
+        yieldTo[_from] = _to;
+        yieldFrom[_to] = _from;
+        rebaseState[_from] = RebaseOptions.YieldDelegationSource;
+        rebaseState[_to] = RebaseOptions.YieldDelegationTarget;
+
+        // Local
+        creditBalances[_from] = fromBalance;
+        alternativeCreditsPerToken[_from] = 1e18;
+        (creditBalances[_to], ) = _balanceToRebasingCredits(
+            fromBalance + toBalance
+        );
+        // Global
+        (
+            uint256 fromCredits,
+            uint256 fromBalanceAccurate
+        ) = _balanceToRebasingCredits(fromBalance);
+        _adjustGlobals(fromCredits.toInt256(), -fromBalanceAccurate.toInt256());
+        emit YieldDelegated(_from, _to);
+    }
+
+    /*
+     * @notice Stop sending the yield from one account to another account.
+     */
+    function undelegateYield(address _from) external onlyGovernor {
+        // Require a delegation, which will also ensure a valid delegation
+        require(yieldTo[_from] != address(0), "Zero address not allowed");
+
+        address to = yieldTo[_from];
+        uint256 fromBalance = balanceOf(_from);
+        uint256 toBalance = balanceOf(to);
+
+        // Remove the bidirectional links
+        yieldFrom[to] = address(0);
+        yieldTo[_from] = address(0);
+        rebaseState[_from] = RebaseOptions.StdNonRebasing;
+        rebaseState[to] = RebaseOptions.StdRebasing;
+
+        // Local
+        // alternativeCreditsPerToken[from] already 1e18 from `delegateYield()`
+        creditBalances[_from] = fromBalance;
+        // alternativeCreditsPerToken[to] already 0 from `delegateYield()`
+        (creditBalances[to], ) = _balanceToRebasingCredits(toBalance);
+        // Global
+        (
+            uint256 fromCredits,
+            uint256 fromBalanceAccurate
+        ) = _balanceToRebasingCredits(fromBalance);
+        _adjustGlobals(
+            -(fromCredits.toInt256()),
```

Comment:
🟥 This results in incorrect accounting at all resolution levels.

When we undelegate, we need to account for new nonRebasingSupply being added, and old credits being removed. The old credits removed is the difference between the big credits that used to be held by the target, and the new smaller credits new held by the target. 

Mathematically, this difference is the `oldTo - newTo = fromCredits`, but that's not the case with rounding, even with high resolution.

```solidity
//setup
ousd.initialize(address(this), 1e26-1);
ousd.mint(nonrebasing, 1000 ether+ 1231245);
ousd.mint(pool, 1000 ether+ 1234345);

ousd.delegateYield(pool, collector);

ousd.changeSupply(ousd.totalSupply()+1e19+1e11+2);
ousd.undelegateYield(pool);
// accounting now messed up
```

Reply:
Corrected this, and other problems around globals in https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/5e57112a9db3bed51d5ef725a589ca89f576df45